staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -39,6 +40,601 @@ func RunTestWatch(ctx context.Context, t *testing.T, store storage.Interface) {
+			},
+			pred: storage.SelectionPredicate{
+				Label: labels.Everything(),
+				Field: fields.SelectorFromSet(fields.Set{
+					"metadata.name": "foo6",
+					"spec.nodeName": "bar4",
+				}),
+				GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+					pod := obj.(*example.Pod)
+					return nil, fields.Set{"spec.nodeName": pod.Spec.NodeName, "metadata.name": pod.ObjectMeta.Name}, nil
+				},
+				IndexFields: []string{"spec.nodeName"},
+			},
+		},
+		{
+			name:          "cluster-wide watch, request with name, field selector with metadata.name, and with spec.nodeName to filter out watch",
+			requestedName: "foo8",
+			watchTests: []*testWatchStruct{
+				{basePod("foo8"), true, watch.Added},
+				{basePod("foo9"), false, ""},
+				{basePodUpdated("foo8"), true, watch.Modified},
+				{basePodAssigned("foo8", "bar5"), true, watch.Deleted},
+				{basePodAssigned("foo9", "bar5"), false, ""},
+			},
+			pred: storage.SelectionPredicate{
+				Label: labels.Everything(),
+				Field: fields.AndSelectors(fields.ParseSelectorOrDie("spec.nodeName!=bar5"), fields.SelectorFromSet(fields.Set{"metadata.name": "foo8"})),
+				GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+					pod := obj.(*example.Pod)
+					return nil, fields.Set{"spec.nodeName": pod.Spec.NodeName, "metadata.name": pod.ObjectMeta.Name}, nil
+				},
+				IndexFields: []string{"spec.nodeName"},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			requestInfo := &genericapirequest.RequestInfo{}
+			requestInfo.Name = tt.requestedName
+			requestInfo.Namespace = ""
+			ctx = genericapirequest.WithRequestInfo(ctx, requestInfo)
+			ctx = genericapirequest.WithNamespace(ctx, "")
+
+			watchKey := "/pods"
+			if tt.requestedName != "" {
+				watchKey += "/" + tt.requestedName
+			}
+
+			list := &example.PodList{}
+			if err := store.GetList(ctx, "/pods", storage.ListOptions{ResourceVersion: "0", Predicate: tt.pred, Recursive: true}, list); err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+
+			w, err := store.Watch(ctx, watchKey, storage.ListOptions{ResourceVersion: list.ResourceVersion, Predicate: tt.pred, Recursive: tt.recursive})
+			if err != nil {
+				t.Fatalf("Watch failed: %v", err)
+			}
+			var prevObj *example.Pod
+			for _, watchTest := range tt.watchTests {
+				out := &example.Pod{}
+				key := "pods/" + watchTest.obj.Name
+				err := store.GuaranteedUpdate(ctx, key, out, true, nil, storage.SimpleUpdate(
+					func(runtime.Object) (runtime.Object, error) {
+						obj := watchTest.obj.DeepCopy()
+						return obj, nil
+					}), nil)
+				if err != nil {
+					t.Fatalf("GuaranteedUpdate failed: %v", err)
+				}
+				expectObj := out
+				if watchTest.expectEvent {
+					if watchTest.watchType == watch.Deleted {
+						expectObj = prevObj
+						expectObj.ResourceVersion = out.ResourceVersion
+					}
+					testCheckResult(t, watchTest.watchType, w, expectObj)
+				}
+				prevObj = out
```

Comment:
Hi @wojtek-t , thank you for pointing out this. I did not find a good way to test filtering watch when there are multiple objects. I just make sure the object is the same with the one just before it when the result channel expects the `watch.Deleted` event. I added comments for those filtering watches just now. Appreciated if you have any suggestion about this.

Reply:
The simplest thing that comes to my mind would be, instead of having `prevObj` have something like:
`
map[string]*example.Pod
`

And us the "namespace/name" as a key for this map.

[Requires making it compile, but high-level sth like
```
currentObj[string]*example.Pod := map[string]*example.Pod{}
for _, watchTest := range tt.watchTests {
   ... // what you have now

    expectObj := out
    podIdentifier := identifier(watchTest.obj)
    if watchTest.watchType == watch.Deleted {
       expectedObj = currentObj[podIdentifier]
       expectedObj.ResourceVersion = out.ResourceVersion
       delete(currentObj, podIdentifier)
    } else {
      currentObj[podIdentifier] = out
    }

    if watchTest.expectEvent {
      testCheckResult(t, watchTest.watchType, w, expectObj)
   }
  
   ... // the remaining part
}
```